neutrinordp/xrdp-neutrinordp.c

```diff
@@ -542,6 +542,52 @@ lxrdp_set_param(struct mod *mod, const char *name, const char *value)
     {
         /* Valid (but unused) parameters not logged */
     }
+    else if (g_strcmp(name, "perf.allow_client_experiencesettings") == 0)
+    {
+            mod->allow_experiencesettings = g_text2bool(value);
+            LOG(LOG_LEVEL_INFO, "%s=[%s] flag:[%d],",
+                             name, value, g_text2bool(value));
+    }
+    else if (g_strcmp(name, "perf.wallpaper") == 0)
+    {
+             strcpy(mod->wallpaper,value);
+             LOG(LOG_LEVEL_INFO, "%s=[%s]",name, value);
+    }
+    else if (g_strcmp(name, "perf.font_smoothing") == 0)
+    {
+             strcpy(mod->font_smoothing,value);
+             LOG(LOG_LEVEL_INFO, "%s=[%s]",name, value);
+    }
+    else if (g_strcmp(name, "perf.desktop_composition") == 0)
+    {
+             strcpy(mod->desktop_composition,value);
+             LOG(LOG_LEVEL_INFO, "%s=[%s]",name, value);
+    }
+    else if (g_strcmp(name, "perf.full_window_drag") == 0)
+    {
+             strcpy(mod->full_window_drag,value);
+             LOG(LOG_LEVEL_INFO, "%s=[%s]",name, value);
+    }
+    else if (g_strcmp(name, "perf.menu_anims") == 0)
+    {
+             strcpy(mod->menu_anims,value);
+             LOG(LOG_LEVEL_INFO, "%s=[%s]",name, value);
+    }
+    else if (g_strcmp(name, "perf.themes") == 0)
+    {
+             strcpy(mod->themes,value);
+             LOG(LOG_LEVEL_INFO, "%s=[%s]",name, value);
+    }
+    else if (g_strcmp(name, "perf.cursor_blink") == 0)
+    {
+             strcpy(mod->cursor_blink,value);
+             LOG(LOG_LEVEL_INFO, "%s=[%s]",name, value);
+    }
+    else if (g_strcmp(name, "perf.cursor_shadow") == 0)
+    {
+             strcpy(mod->cursor_shadow,value);
+             LOG(LOG_LEVEL_INFO, "%s=[%s]",name, value);
+    }
```

Comment:
This comment applies to many sections below.

There's a lot of lines here which aren't really necessary.

Firstly, there's a C idiom that you should be aware of to clear bits in a word. You're doing it by setting the bits and then using XOR to change the value. That will definitely work, but it's not the standard way to do it, which is as follows:-

Standard way to set bits in a word:-
```c
mod->value |= bitmask
```

Standard way to clear bits in a word:-
```c
mod->value &= ~bitmask
```

More info [here](https://stackoverflow.com/questions/3920307) [stackoverflow.com]

With that out the way, I can show you a way to implement your change which needs quite a bit less code, and avoids string copies. Unnecessary string copies in C are the source of many problems.

Rather than storing a string for each parameter, all you need are two words in `struct mod` in `xrdp-neutrinordp.h`:-

```
    int perf_settings_override_mask; /* Performance bits overridden in ini file */
    int per_settings_values_mask; /* Values of overridden performance bits */
```

- The override mask has a bit set for every value that the user has overridden, regardless of the value.
- The value mask has the values of the bits which are actually set in the override mask

Then in the code above, your settings become something like this. Note the difference in the code between a disable setting (PERF_DISABLE_WALLPAPER) and an enable setting (PERF_ENABLE_FONT_SMOOTHING)

```c
    /* Initialise performance settings masks */
    mod->perf_settings_override_mask = 0;
    mod->per_settings_values_mask = 0;
    . . .
    else if (g_strcmp(name, "perf.wallpaper") == 0)
    {
        mod->perf_settings_override_mask |= PERF_DISABLE_WALLPAPER;
        if (!g_text2bool(value))
        {
            mod->per_settings_values_mask |= PERF_DISABLE_WALLPAPER;
        }
    }
    else if (g_strcmp(name, "perf.font_smoothing") == 0)
    {
        mod->perf_settings_override_mask |= PERF_ENABLE_FONT_SMOOTHING;
        if (g_text2bool(value))
        {
            mod->per_settings_values_mask |= PERF_ENABLE_FONT_SMOOTHING;
        }
    }
    . . .
```

Then below, your lines 1770 - 1927 can be simply replaced with:-

```c
/* Clear bits for any overridden performance settings */
instance->settings->performance_flags =
    (instance->settings_performance_flags & ~mod->perf_settings_override_mask);

/* Add in overridden performance settings */
instance->settings->performance_flags |= mod->perf_settings_values_mask;
```

Does that make sense?

Reply:
Hi @matt335672 

Thank you for your very detailed advice on my verbose and immature code.

I appreciate your guidance on bit operations. I will relearn it.

Thank you also for the workaround for copying strings.
When the parameter was false and when I commented it out, the variable value became the same, so I used the method of judging by string.

I will fix the code based on the advice you gave.
It will take some time.

Reply:
Please feel free to ask any questions if you think I haven't been clear enough, and thanks for doing this.